pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -1302,6 +1312,68 @@ public CompletableFuture<Void> updateSubscriptionProperties(Map<String, String>
                     this.subscriptionProperties = newSubscriptionProperties;
                 });
     }
+
+    public void updateFilterStats() {
+        Dispatcher dispatcher0 = getDispatcher();
+        if (dispatcher0 == null) {
+            // dispatcher is not initialized yet
+            return;
+        }
+        double oldFilterAcceptedRateEstimation = this.filterAcceptedRateEstimation;
+
+        // update the estimated filter accepted rate
+        long processed = dispatcher0.getFilterProcessedMsgCount();
+        long accepted = dispatcher0.getFilterAcceptedMsgCount();
+
+        if (oldFilterAcceptedRateEstimation == 1.0 && processed == accepted) {
+            // filter is accepting all the messages
+            return;
+        }
+
+        double currentFilterAcceptedRateEstimation = 1.0;
+        if (processed != accepted && processed > 0) {
+            currentFilterAcceptedRateEstimation = accepted * 1.0 / processed;
+        }
+
+        //TODO: take time into consideration
+        this.filterAcceptedRateEstimation = (oldFilterAcceptedRateEstimation + currentFilterAcceptedRateEstimation) / 2;
+
+        if (oldFilterAcceptedRateEstimation == filterAcceptedRateEstimation) {
+            // no need to write to metadata store
+            return;
+        }
+
+        log.info("Persist new filter stats for subscription {} on topic {} : filterAcceptedRateEstimation={}",
+                getName(), getTopic().getName(), currentFilterAcceptedRateEstimation);
+
+        Map<String, String> subscriptionProperties = new HashMap<>(getSubscriptionProperties());
+        subscriptionProperties.put("filterAcceptedRateEstimation",
+                Double.toString(currentFilterAcceptedRateEstimation));
+        updateSubscriptionProperties(subscriptionProperties).whenComplete((ignore, ex) -> {
+            if (ex != null) {
+                log.warn("Failed to persist filter stats for subscription {} on topic {}",
+                        getName(), getTopic().getName(), ex);
+            }
+        });
```

Comment:
I looked into it, and if we wanted to store this metadata in bookkeeper, we'd need to expand the `ManagedCursorInfo` protobuf object. Putting it in zk seems like a good initial solution, and if we find that the load is too high, we have options.

Reply:
yes, ZK is what we have at the moment, it is the less disruptive change

Reply:
I agree, I just wanted to follow up on it because I had brought it up in the google doc.